internal/mining/mining_view.go

```diff
@@ -0,0 +1,464 @@
+func (mv *TxMiningView) updateStatsDescendantsRemoved(baseTxDesc *TxDesc) {
+	baseTxHash := baseTxDesc.Tx.Hash()
+	if _, hasStats := mv.ancestorStats[*baseTxHash]; !hasStats {
+		// If the transaction does not have ancestor tracking enabled, then
+		// none of its descendants should either.
+		return
+	}
+
+	numUntrackedDescendants := 0
+	seenDescendants := make(map[chainhash.Hash]struct{}, AncestorTrackingLimit)
+	mv.txGraph.forEachDescendantPreOrder(baseTxHash, seenDescendants,
+		func(descendantTxDesc *TxDesc) bool {
+			descendantTxHash := *descendantTxDesc.Tx.Hash()
+			descendantStats, hasStats := mv.ancestorStats[descendantTxHash]
+
+			// Attempt to track the descendant's ancestor stats if it was not
+			// tracked previously. This scenario applies to a transaction
+			// that once had too many ancestors, but is now eligible to
+			// have stats tracked.
+			// A check is also performed to avoid testing too many descendants
+			// for eligibility to be included in the tracked set, since no
+			// descendant can know if any one of its ancestors has too many
+			// descendants tracked without visiting the ancestor. Note that
+			// although this has a non-ideal behavior of potentially skipping
+			// transactions that would otherwise be eligible to have stats
+			// tracked, it bounds the number of times ancestors are walked.
+			if !hasStats && numUntrackedDescendants < AncestorTrackingLimit {
+				numUntrackedDescendants++
+
+				// Note that this call retrieves ancestors while walking
+				// descendants of another transaction, and that the time
+				// complexity is bound by the maximum number of ancestors
+				// allowed to be tracked in the graph and the maximum
+				// number of descendants allowed to have ancestor stats tracked.
+				mv.maybeUpdateAncestorStats(descendantTxDesc, baseTxHash)
+
+				// Do not walk descendants of this descendant if it did not
+				// have ancestor stats previously because it was on the edge of
+				// the ancestor limit. Any transactions that descend from it
+				// would exceed the limit by at least one.
+				return false
+			}
+
+			if hasStats {
+				removeAncestorFrom(descendantStats, baseTxDesc)
+				return true
+			}
+
+			// If the transaction does not have statistics tracked then none of
+			// its descendants will either.
+			return false
+		})
+
+	// Update all ancestors to account for the removal of a descendant.
+	seenAncestors := make(map[chainhash.Hash]*TxDesc, AncestorTrackingLimit)
+	mv.txGraph.forEachAncestorPreOrder(baseTxHash, seenAncestors,
+		func(ancestorTxDesc *TxDesc) bool {
+			ancestorStats, exists := mv.ancestorStats[*ancestorTxDesc.Tx.Hash()]
+			if exists {
+				removeDescendantFrom(ancestorStats)
+			}
+			return true
+		})
+}
+
+// AddTransaction inserts a TxDesc into the mining view if it has a parent
+// or child in the view, or has a relationship with another transaction through
+// findTx.
+func (mv *TxMiningView) AddTransaction(txDesc *TxDesc, findTx txDescFind) {
```

Comment:
Same thing here regarding the unexported type.

Reply:
I exported the type for this case, it's being used in quite a few places so it seems to make sense to have the type defined to me.